--- a/zones.c
+++ b/zones.c
@@ -502,12 +502,19 @@ print_zone_rule(struct fw3_ipt_handle *h
 		// these should go here specifically, after all the other firewall forwarding rules but before the default rule
 		if (has(zone->flags, handle->family, FW3_FLAG_DREAMCATCHER))
 		{
-      // First rule sends packets to DREAMCATCHER_lan_rule chain
+      // First rule sends mDNS packets straight to NFQUEUE, skipping the DREAMCATCHER_lan_rule chain
+      r = fw3_ipt_rule_new(handle);
+			fw3_ipt_rule_extra(r, "-p UDP --dport 5353");
+      fw3_ipt_rule_comment(r,"mDNS -> NFQUEUE");
+      fw3_ipt_rule_target(r, "NFQUEUE");
+      fw3_ipt_rule_addarg(r, NULL, "--queue-num", "4670"); // TODO: make the queue number configurable
+      fw3_ipt_rule_append(r, "zone_%s_forward", zone->name);
+      // Second rule sends packets to DREAMCATCHER_lan_rule chain
 			r = fw3_ipt_rule_new(handle);
 			fw3_ipt_rule_comment(r, "DREAMCATCHER chain");
 			fw3_ipt_rule_target(r, "DREAMCATCHER_%s_rule", zone->name);
 			fw3_ipt_rule_append(r, "zone_%s_forward", zone->name);
-      // Second rule sends packets to NFQUEUE if they didn't match any existing rules
+      // Third rule sends packets to NFQUEUE if they didn't match any existing rules
       r = fw3_ipt_rule_new(handle);
       fw3_ipt_rule_comment(r,"Default DREAMCATCHER -> NFQUEUE");
       fw3_ipt_rule_target(r, "NFQUEUE");
