diff --git a/package/system/ca-certificates/Makefile b/package/system/ca-certificates/Makefile
index 9b6dbd61fe..168cb0dfea 100644
--- a/package/system/ca-certificates/Makefile
+++ b/package/system/ca-certificates/Makefile
@@ -7,12 +7,12 @@
 include $(TOPDIR)/rules.mk
 
 PKG_NAME:=ca-certificates
-PKG_VERSION:=20160104
+PKG_VERSION:=20151214
 PKG_MAINTAINER:=Christian Schoenebeck <christian.schoenebeck@gmail.com>
 
 PKG_SOURCE:=$(PKG_NAME)_$(PKG_VERSION).tar.xz
 PKG_SOURCE_URL:=http://ftp.debian.org/debian/pool/main/c/ca-certificates
-PKG_MD5SUM:=d9665a83d0d3ef8176a38e6aa20458e9
+PKG_MD5SUM:=2233bfa64af6f58f5eca9735b6742818
 PKG_BUILD_DIR:=$(BUILD_DIR)/$(PKG_NAME)
 
 PKG_INSTALL:=1
